Add Azure IoT Hub and FOTA libraries and samples

jtguggedal opened the pull request:
This PR adds the following
- Azure FOTA library and sample
- Azure IoT Hub library and sample

I will add documentation in separate commits and squash into the corresponding library or sample commits.

Depends on #2750

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:136: WARNING:LONG_LINE_COMMENT: line over 80 characters
#136: FILE: include/net/azure_iot_hub.h:43:
+	/** Device twin update successful. The request ID and status are contained

-:700: WARNING:LINE_SPACING: Missing a blank line after declarations
#700: FILE: samples/nrf9160/azure_fota/src/main.c:163:
+		int err = lte_lc_init_and_connect_async(lte_handler);
+		if (err) {

-:1099: WARNING:LONG_LINE_STRING: line over 80 characters
#1099: FILE: samples/nrf9160/azure_iot_hub/src/main.c:69:
+	event_interval_obj = cJSON_GetObjectItem(desired_obj, "telemetryInterval");

-:1266: ERROR:CODE_INDENT: code indent should use tabs where possible
#1266: FILE: samples/nrf9160/azure_iot_hub/src/main.c:236:
+^I^I        atomic_get(&event_interval));$

-:1277: WARNING:STATIC_CONST_CHAR_ARRAY: static char array declaration should probably be static const char
#1277: FILE: samples/nrf9160/azure_iot_hub/src/main.c:247:
+	static char ret[] = "{\"this\":\"worked\"}";

-:1413: WARNING:LINE_SPACING: Missing a blank line after declarations
#1413: FILE: samples/nrf9160/azure_iot_hub/src/main.c:383:
+		int err = lte_lc_init_and_connect_async(lte_handler);
+		if (err) {

-:2076: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#2076: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:28:
+#define USER_NAME_STATIC ^ICONFIG_AZURE_IOT_HUB_HOSTNAME "/"^I\$

-:2080: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#2080: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:32:
+ * ^I<IoT Hub hostname>/<device ID>/?api-version=2018-06-30$

-:2084: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#2084: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:36:
+#define TOPIC_DEVICEBOUND ^I"devices/%s/messages/devicebound/#"$

-:2088: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#2088: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:40:
+#define TOPIC_EVENTS^I ^I"devices/%s/messages/events/"$

-:2120: WARNING:BLOCK_COMMENT_STYLE: Block comments use a trailing */ on a separate line
#2120: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:72:
+ * ".azure-devices-provisioning.net/" and "/?api-version=2018-06-30" */

-:2122: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#2122: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:74:
+#define USER_NAME_BUF_LEN ^ICONFIG_AZURE_IOT_HUB_NAME_MAX_LEN + \$

-:2122: WARNING:COMPLEX_MACRO: Macros with complex values should be enclosed in parentheses
#2122: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:74:
+#define USER_NAME_BUF_LEN 	CONFIG_AZURE_IOT_HUB_NAME_MAX_LEN + \
+				CONFIG_AZURE_IOT_HUB_DEVICE_ID_MAX_LEN + \
+				sizeof(".azure-devices-provisioning.net/") + \
+				sizeof("/?api-version=2018-06-30")

-:2392: WARNING:BLOCK_COMMENT_STYLE: Block comments should align the * on each line
#2392: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:344:
+	/* Status codes
+		*	200: Response to request for device twin from

-:2751: ERROR:INITIALISED_STATIC: do not initialise statics to false
#2751: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:703:
+	static bool certs_added = false;

-:2863: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#2863: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:815:
+^I^Itls_cfg->hostname ^I= CONFIG_AZURE_IOT_HUB_DPS_HOSTNAME;$

-:2995: WARNING:TYPO_SPELLING: 'progess' may be misspelled - perhaps 'progress'?
#2995: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:947:
+static int fota_progess_report(uint32_t progress)

-:3007: WARNING:SUSPECT_CODE_INDENT: suspect code indent for conditional statements (8, 20)
#3007: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:959:
+	if (progress > AZURE_FOTA_EVT_DL_COMPLETE_VAL ||
[...]
+		    LOG_DBG("Invalid progress value: %d", progress);

-:3010: WARNING:TABSTOP: Statements should start on a tabstop
#3010: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:962:
+		    return -EINVAL;

-:3065: WARNING:TYPO_SPELLING: 'progess' may be misspelled - perhaps 'progress'?
#3065: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:1017:
+		fota_progess_report(0);

-:3089: WARNING:TYPO_SPELLING: 'progess' may be misspelled - perhaps 'progress'?
#3089: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:1041:
+		fota_progess_report(fota_evt->dl.progress);

-:3124: ERROR:OPEN_BRACE: that open brace { should be on the previous line
#3124: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:1076:
+	switch (tx_data->topic.type)
+	{

-:3382: WARNING:MISSING_BREAK: Possible switch case/default not preceded by break or fallthrough comment
#3382: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub_cloud.c:48:
+	case AZURE_IOT_HUB_EVT_FOTA_START:

-:3670: WARNING:TYPO_SPELLING: 'eror' may be misspelled - perhaps 'error'?
#3670: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:144:
+		LOG_ERR("Failed to initialize settings subsystem, eror: %d",

-:3749: WARNING:ONE_SEMICOLON: Statements terminations use 1 semicolon
#3749: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:223:
+	root_obj = cJSON_Parse(json);;

-:3784: WARNING:ONE_SEMICOLON: Statements terminations use 1 semicolon
#3784: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:258:
+	root_obj = cJSON_Parse(json);;

-:3871: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#3871: FILE: subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:345:
+^I * ^I200: Device registration has been processed$

- total: 3 errors, 24 warnings, 3984 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPDX_LICENSE_TAG SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.
```

## Nits issues

```
Please remove blank lines at end of 'samples/nrf9160/azure_fota/src/main.c'

Please remove blank lines at end of 'subsys/net/lib/azure_fota/azure_fota.c'
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

coderbyheart commented on subsys/net/lib/azure_fota/azure_fota.c:
```diff
@@ -0,0 +1,277 @@
+	/* Desired object */
+	desired_obj = cJSON_GetObjectItemCaseSensitive(root_obj, "desired");
+	if (desired_obj == NULL) {
+		/* The received object can be the 'desired' object only, or
+		 * the whole device twin properties object. If there is no
+		 * 'desired' object to be found, assume it's the root object.
+		 */
+		LOG_DBG("No 'desired' object found, assuming it's root object");
+		desired_obj = root_obj;
```
Storing data on the Device Twin root is an undocumented behaviour: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-device-twins ... so this case should not be handled.

jtguggedal replied:
The "root" in this case refers to the root of the received object and not the device twin per se. When the device twin `desired` object is updated, only the `desired` object is received on `$iothub/twin/PATCH/properties/desired/#`, while if the device twin is explicitly requested, the `properties` object of the device twin is received on `$iothub/twin/res/#`. So that's why we need to look for the `firmware` object in the `desired` object it that exists, or in the root of the received object if `desired` doesn't exist.

This should probably be made clearer in the comment in the code, and possibly also have a parameter in `azure_fota_msg_process` to indicate where to look for the `firmware` object.

coderbyheart replied:
Ah, after double checking the comment I see that it says that.

coderbyheart commented on subsys/net/lib/azure_fota/azure_fota.c:
```diff
@@ -0,0 +1,277 @@
+		goto clean_exit;
+	}
+
+	/* TODO: Do something sensible with this check value */
```
The [official guide](https://docs.microsoft.com/en-us/azure/iot-hub/tutorial-firmware-update) does not provide any recommendation of what to do with the `fwCheckValue` field.

Looking at [some sample IoT clients](https://github.com/search?q=org%3AAzure-Samples+fwPackageCheckValue&type=Code) this should contain the checksum of the firmware file which the device can use to verify that it downloaded the file correctly.

We don't do this [in the AWS FOTA](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/net/aws_fota.html) either because the integrity check is part of the FW file itself and the modem handles that.

So I recommend to drop it.

jtguggedal replied:
Yeah, sounds reasonable to drop that field, as you say, it will be checked anyway by the respective bootloader receiving the image.

coderbyheart commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1348 @@
+#define DPS_USER_NAME		CONFIG_AZURE_IOT_HUB_DPS_ID_SCOPE \
+				"/registrations/%s/api-version=2019-03-31"
+
+#define FOTA_PROGRESS_REPORT	"{\"firmware\":{\"status\":\"%s\",\"dl\":%d}}"
```
Could we add the `fwVersion` property here, which is copied from the same property on the desired object? Otherwise it is not clear *which* firmware the device is installing. Because it is not guaranteed that the device always picks up the latest FOTA job.

jtguggedal replied:
Agreed. The current reporting mechanism works more like a placeholder for testing purposes.
The firmware version currently being downloaded should definitely be reported. It should also report back after the device has rebooted and actually applied the new firmware image to indicate whether that succeeded or not, ie implement the rest of the [recommended ](https://docs.microsoft.com/en-us/azure/iot-hub/tutorial-firmware-update) `fwUpdateStatus` states, along with `currentFwVersion`, `pendingFwVersion` and possibly `fwUpdateSubstatus`.

coderbyheart replied:
To my understanding this is already solved:
1. when downloading the firmware, the device reports the **new** version in `pendingFwVersion` before rebooting
2. after reboot 
   1. if the update was **successful** the application will report its current version (which is hardcoded in the firmware) in `currentFwVersion` AND `pendingFwVersion`; they will now contain the *new* version.
   2. if the update was **not successful** `currentFwVersion` AND `pendingFwVersion` will then contain the *old* version.

coderbyheart commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1348 @@
+#define DPS_USER_NAME		CONFIG_AZURE_IOT_HUB_DPS_ID_SCOPE \
+				"/registrations/%s/api-version=2019-03-31"
+
+#define FOTA_PROGRESS_REPORT	"{\"firmware\":{\"status\":\"%s\",\"dl\":%d}}"
```
`dl` propery should only be included if download progress reporting is enabled.

jtguggedal replied:
Yep. I thought I had added a Kconfig option to leave this out altogether, but apparently I didn't.

coderbyheart replied:
Maybe I have not read the source correctly, when this is used.
I assumed it was used also to report start and finish of a FOTA.

jtguggedal replied:
It's used for that, yes. I meant that I thought I had added the option to leave out the `dl` field and only report states like `downloading` and `applying`, but seems `dl` is always added, and the hardcoded value on line 49 makes the download process being sent on every 20% of progress, which doesn't add a whole lot of value. 

I'll make if configurable whether progress reports shall be sent at all, and if so, at which progress increments [%] it should be sent.

coderbyheart commented on subsys/net/lib/azure_fota/azure_fota.c:
```diff
@@ -0,0 +1,277 @@
+	 *		fwLocation : {
+	 *			host : <hostname>,
+	 *			path : <file path>
+	 *		},
```
Is this still needed? The pre-parsing was introduced so that the device could download from AWS S3 using a non-secure URL (it would construct the non-https url using the properties in the struct).

See https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/net/aws_fota.html#limitations

And either way, I would prefer to have a string value here, because in both case it's the users responsibility to provide a non-secure download location.

coderbyheart replied:
`fwPackageURI` is the property used in [the reference](https://docs.microsoft.com/en-us/azure/iot-hub/tutorial-firmware-update#update-the-firmware)

jtguggedal replied:
Do you mean why it's split into `host`and `path`? It's just convenience, as that's the format the FOTA download client in NCS requires to have it (it can download multiple images from the same host and apply it to different partitions). I decided to let the cloud side take the job of splitting into host and path instead of parsing it on the device.

coderbyheart replied:
Alright.

jtguggedal replied:
> I can change the name into fwPackageURI, although host and path would probably also have to change names from what I read here.

Or maybe not, I read the example wrong.

coderbyheart replied:
Let's keep it `fwLocation` because `fwPackageURI` in the Azure tutorial is a string, so I'd rather not create an overlap of different types.

Also, it's not a full URI, because `scheme` is omitted, because it's hardcoded to `http`.

coderbyheart replied:
In Bifravst I have implemented it so that the download URL is stored as string in `fwPackageURI` and parsed in `fwLocation`: https://github.com/bifravst/azure/blob/79a30865eef9c2bbf7c4b301171384a1b5ddb6f8/features/FOTA.feature#L43-L69

simensrostad commented on subsys/net/lib/azure_iot_hub/Kconfig:
```diff
@@ -0,0 +1,104 @@
+config AZURE_IOT_HUB_DEVICE_ID
+	string "Device ID"
+	help
+	  Device ID to be used when connecting to IoT Hub and aptionally DPS.
```
aptionally

simensrostad commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_dps.h:
```diff
@@ -0,0 +1,78 @@
+ *	    connect to an IoT hub.
+ */
+int dps_start(void);
+
```
Undocumented functions

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_cloud.c:
```diff
@@ -0,0 +1,187 @@
+	struct cloud_backend_config *config = azure_iot_hub_backend->config;
+	struct cloud_event cloud_evt = { 0 };
+
+	switch (evt->type) {
```
translation of AZURE_IOT_HUB_EVT_CONNECTING to CLOUD.....EVT_CONNECTING is missing.

jtguggedal replied:
Nice catch

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:
```diff
@@ -0,0 +1,612 @@
+
+	status_item =
+		cJSON_GetObjectItem(reg_status_obj, "status");
+	if (reg_status_obj == NULL) {
```
Here the return value should be checked?

jtguggedal replied:
Good catch, fixed

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:
```diff
@@ -0,0 +1,612 @@
+
+	assigned_hub_item =
+		cJSON_GetObjectItem(reg_status_obj, "assignedHub");
+	if (reg_status_obj == NULL) {
```
Also here

simensrostad reviewed (commented):
First glance, a lot of nit.

jtguggedal commented:
I've added initial documentation for the FOTA library and sample, and addressed the comments regarding FOTA.

coderbyheart commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,448 @@
+void main(void)
+{
+	int err;
+	/* The device id below is only used in byt the library if
```
```suggestion
	/* The device ID below is only used in by the library if
```

coderbyheart commented on subsys/net/lib/azure_fota/azure_fota.c:
```diff
@@ -0,0 +1,565 @@
+		       strlen(get_current_version())) != 0;
+}
+
+/* Static functions */
```
```suggestion
```

Obvious.

coderbyheart commented on subsys/net/lib/azure_fota/azure_fota.c:
```diff
@@ -0,0 +1,565 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
```
Should contain a link to https://docs.microsoft.com/en-us/azure/iot-hub/tutorial-firmware-update because this implementation is based on the process outlined in that document.

coderbyheart commented on subsys/net/lib/azure_iot_hub/Kconfig:
```diff
@@ -0,0 +1,104 @@
+if AZURE_IOT_HUB
+
+config AZURE_IOT_HUB_NAME
+	string "Azure IoT Hub name"
```
Please provide example of what a `name` looks like.

coderbyheart commented on subsys/net/lib/azure_iot_hub/Kconfig:
```diff
@@ -0,0 +1,104 @@
+	default 2560
+
+config AZURE_IOT_HUB_AUTO_DEVICE_TWIN_REQUEST
+	bool "Request device automatically when connected"
```
```suggestion
	bool "Request device twin automatically when connected"
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_dps.h:
```diff
@@ -0,0 +1,78 @@
+int dps_init(struct dps_config *cfg);
+
+/* @brief Parse incoming MQTT message to see if it's DPS related and process
+ *	  accordingly if it is.
```
```suggestion
 *	  accordingly.
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_dps.h:
```diff
@@ -0,0 +1,78 @@
+ *	  accordingly if it is.
+ *
+ * @retval true The message was DPS-related and is consumed.
+ * @retval false The message has not DPS-related.
```
```suggestion
 * @retval false The message was not DPS-related.
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_dps.h:
```diff
@@ -0,0 +1,78 @@
+ */
+bool dps_process_message(struct azure_iot_hub_evt *evt);
+
+/* @brief Start Azure device provisioning service.
```
```suggestion
/* @brief Start Azure Device Provisioning Service process.
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1322 @@
+
+static azure_iot_hub_evt_handler_t evt_handler;
+
+/* If DPS is used, the IoT hub name is obtained through that service, otherwise
```
```suggestion
/* If DPS is used, the IoT Hub hostname is obtained through that service, otherwise
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_cloud.c:
```diff
@@ -0,0 +1,187 @@
+	err = azure_iot_hub_connect();
+	if (err >= 0) {
+		/* At this point, we can't be sure that the socket will be
+		 * used for the IoT hub connection or the DPS connection only.
```
```suggestion
		 * used for the IoT Hub connection or the DPS connection only.
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+1. `Download nRF Connect for Desktop`_
+#. Update the modem firmware on the on-board modem of the nRF9160 based board to the latest version by following the steps in `Updating the nRF9160 DK cellular modem`_.
+#. Build and program the  :ref:`at_client_sample` sample to the nRF9160 based board as explained in :ref:`gs_programming`.
+#. Launch the `LTE Link Monitor`_ application, which is implemented as part of `nRF Connect for Desktop`_.
```
```suggestion
#. Launch the `LTE Link Monitor`_ application, which part of `nRF Connect for Desktop`_.
```

It's a separate application in there.

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+#. Launch the `LTE Link Monitor`_ application, which is implemented as part of `nRF Connect for Desktop`_.
+#. Click **Certificate manager** located at the top right corner.
+#. Copy the Baltimore CyberTrust Root certificate into the ``CA certificate`` entry.
+#. Copy and paste the device certificate and key created using the scripts found at `Create Azure IoT Hub certificates`, into the respective entries (``Client certificate``, ``Private key``).
```
```suggestion
#. Copy and paste the device certificate and key created using the scripts found at `Create Azure IoT Hub certificates`_, into the respective entries (``Client certificate``, ``Private key``).
```

use link again

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+#. Click **Certificate manager** located at the top right corner.
+#. Copy the Baltimore CyberTrust Root certificate into the ``CA certificate`` entry.
+#. Copy and paste the device certificate and key created using the scripts found at `Create Azure IoT Hub certificates`, into the respective entries (``Client certificate``, ``Private key``).
+#. Select a desired security tag and click **Update certificates**.
```
```suggestion
#. Select a desired security tag (can be any positive integer, e.g. ``42``) and click **Update certificates**.
```

Give example for Sec Tag

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+#. Select a desired security tag and click **Update certificates**.
+
+.. note::
+   The chosen security tag when provisioning the certificates must be equal to the security configured by :option:`CONFIG_AZURE_IOT_HUB_SEC_TAG`.
```
```suggestion
   The chosen security tag when provisioning the certificates must be equal to the security tag configured by :option:`CONFIG_AZURE_IOT_HUB_SEC_TAG`.
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+************
+
+The library is initialized by calling the :cpp:func:`azure_iot_hub_init` function.
+If this API fails, the application can not use any APIs of the library.
```
```suggestion
If the initialization fails, the application can not use any APIs of the library.
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+- ``$iothub/twin/res/#``
+- ``$iothub/methods/POST/#``
+
+Currently, the library does not persistent MQTT sessions enabled, so subscriptions are requested on each connection to the IoT Hub.
```
```suggestion
Currently, the library does not persistent MQTT sessions, so subscriptions are requested on each connection to the IoT Hub.
```

coderbyheart commented on doc/nrf/links.txt:
```diff
@@ -227,6 +227,15 @@
+.. _`Azure Portal`: https://portal.azure.com/
+.. _`Create Azure IoT Hub certificates`: https://github.com/Azure/azure-iot-sdk-c/blob/master/tools/CACertificates/CACertificateOverview.md
+.. _`Azure IoT Hub direct method`: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-direct-methods
+.. _`Baltimore CyberTrust Root certificate`: https://www.digicert.com/CACerts/BaltimoreCyberTrustRoot.crt.pem
```
```suggestion
.. _`Azure IoT Hub MQTT protocol support`: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support
```

(used below)

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+The initial TLS handshake takes some time to complete, usually in the range of seconds up to the low tens of seconds, depending on the network conditions.
+For the duration of the TLS handshake, :cpp:func:`azure_iot_hub_connect` blocks, so care must be taken when deciding from which context the API is called.
+
+After a successful connection, the library automatically subscribes to the following standard Azure IoT Hub MQTT topics:
```
```suggestion
After a successful connection, the library automatically subscribes to the following standard Azure IoT Hub MQTT topics (see `Azure IoT Hub MQTT protocol support`_ for details):
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+For the duration of the TLS handshake, :cpp:func:`azure_iot_hub_connect` blocks, so care must be taken when deciding from which context the API is called.
+
+After a successful connection, the library automatically subscribes to the following standard Azure IoT Hub MQTT topics:
+- ``devices/<device ID>/messages/devicebound/#``
```
```suggestion
- ``devices/<device ID>/messages/devicebound/#`` (cloud-to-device messages)
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+
+After a successful connection, the library automatically subscribes to the following standard Azure IoT Hub MQTT topics:
+- ``devices/<device ID>/messages/devicebound/#``
+- ``$iothub/twin/PATCH/properties/desired/#``
```
```suggestion
- ``$iothub/twin/PATCH/properties/desired/#`` (desired properties update notifications)
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+- ``devices/<device ID>/messages/devicebound/#``
+- ``$iothub/twin/PATCH/properties/desired/#``
+- ``$iothub/twin/res/#``
+- ``$iothub/methods/POST/#``
```
```suggestion
- ``$iothub/methods/POST/#`` (direct method requests)
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,124 @@
+After a successful connection, the library automatically subscribes to the following standard Azure IoT Hub MQTT topics:
+- ``devices/<device ID>/messages/devicebound/#``
+- ``$iothub/twin/PATCH/properties/desired/#``
+- ``$iothub/twin/res/#``
```
```suggestion
- ``$iothub/twin/res/#`` (operation responses)
```

jtguggedal commented on samples/nrf9160/azure_iot_hub/README.rst:
```diff
@@ -0,0 +1,135 @@
+.. include:: /includes/build_and_run.txt
+.. include:: /includes/spm.txt
+
+Testing
```
Link to the library docs with explanations has been added now to the "Requirements"  section

jtguggedal replied:
Should be fixed now

jtguggedal commented on samples/nrf9160/azure_iot_hub/certs/certificates.h:
```diff
@@ -0,0 +1,17 @@
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+static const unsigned char ca_certificate[] = {
```
Certificates in the FOTA sample have been removed as it wasn't fully implemented. So I'll keep this as is and use it as a template when/if adding certificate provisioning to the FOTA sample

jtguggedal commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,432 @@
+ * device twin document.
+ */
+#define EVENT_INTERVAL		20
+#define RECV_BUF_SIZE		300
```
Should be fixed, at least locally it's fine

jtguggedal commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,432 @@
+		break;
+	case AZURE_IOT_HUB_EVT_TWIN:
+		printk("AZURE_IOT_HUB_EVT_TWIN\n");
+		event_interval_apply(event_interval_get(evt->data.msg.ptr));
```
Added `on_evt_*()` functions to keep the switch cleaner and more readable

jtguggedal commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,432 @@
+		printk("azure_iot_hub_connect failed: %d\n", err);
+		return;
+	}
+}
```
Yeah, I can do that

jtguggedal commented on samples/nrf9160/azure_fota/README.rst:
```diff
@@ -0,0 +1,161 @@
+
+   If HTTPS is used as transport layer, certificates must be provisioned to the device and the sec tag option must be set accordingly.
+
+.. option:: AZURE_IOT_HUB_NAME - Azure IoT Hub name.
```
The IoT Hub name was used in the early stages of development instead of the full hostname. I've moved to using the full hostname name and removed remnants of the IoT Hub name usage.

jtguggedal commented on samples/nrf9160/azure_fota/README.rst:
```diff
@@ -0,0 +1,161 @@
+Testing
+=======
+
+After programming the sample to the board, test it by performing the following steps:
```
Explanations added

jtguggedal commented on samples/nrf9160/azure_iot_hub/README.rst:
```diff
@@ -0,0 +1,135 @@
+An alternative to set the device ID in Kconfig, is to enable :option:`CONFIG_AZURE_IOT_HUB_DEVICE_ID_APP` and set the device ID runtime in the configuration struct passed to :cpp:func:`azure_iot_hub_init`.
+
+
+.. option:: :option:`CONFIG_AZURE_IOT_HUB_NAME`
```
Same as above, hub name was earlier used to construct the hostname, but's now removed

jtguggedal commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,448 @@
+	}
+}
+
+/* For simplicity the functions below that send data to cloud are in this
```
Reworked to use a dedicated work queue for potentially blocking socket calls, and the confusing comments have been omitted altogether

jtguggedal commented on samples/nrf9160/azure_fota/prj.conf:
```diff
@@ -0,0 +1,93 @@
+CONFIG_AZURE_IOT_HUB_PORT=8883
+CONFIG_AZURE_IOT_HUB_SEC_TAG=
+CONFIG_AZURE_IOT_HUB_LOG_LEVEL_DBG=y
+# CONFIG_AZURE_IOT_HUB_DPS=y
```
Nope. Removed, and a comment about the option has been added

jtguggedal commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,432 @@
+		break;
+	case AZURE_IOT_HUB_EVT_READY:
+		printk("AZURE_IOT_HUB_EVT_READY\n");
+		k_delayed_work_submit(&send_event_work, K_SECONDS(3));
```
Can't recall... Changed to `K_NO_WAIT`

jtguggedal commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:
```diff
@@ -0,0 +1,612 @@
+
+#define DPS_REG_STATUS_UPDATE_MAX_RETRIES	10
+#define DPS_OPERATION_ID_MAX_LEN		60
+#define DPS_TOPIC_OPERATION_ID_MAX_LEN		(DPS_OPERATION_ID_MAX_LEN + 100)
```
Looks okay locally

jtguggedal commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,283 @@
+enum azure_iot_hub_evt_type {
+	/** Connecting to Azure IoT Hub. */
+	AZURE_IOT_HUB_EVT_CONNECTING = 0x1,
+	/** Connected to Azure IoT Hub MQTT broker. */
```
Changed to `/** Connected to Azure IoT Hub. */`, I think it's sufficient

jtguggedal commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,283 @@
+	/** Acknowledgement for data sent to IoT Hub. */
+	AZURE_IOT_HUB_EVT_PUBACK,
+	/** Device twin has been received. */
+	AZURE_IOT_HUB_EVT_TWIN,
```
Good point, `_RECEIVED` suffix added

jtguggedal commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_cloud.c:
```diff
@@ -0,0 +1,187 @@
+		cloud_evt.type = CLOUD_EVT_FOTA_ERASE_DONE;
+		cloud_notify_event(azure_iot_hub_backend, &cloud_evt,
+				   config->user_data);
+		break;
```
Yeah, I chose not to do that. It's mostly noise with these events

jtguggedal commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,283 @@
+ */
+struct azure_iot_hub_result {
+	/** Request ID to which the result belongs. */
+	uint32_t rid;
```
Possibly. I kept them separate for now for clarity, but I'm open to change this

jtguggedal commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,283 @@
+ */
+int azure_iot_hub_method_respond(struct azure_iot_hub_result *result);
+
+/** @brief Get the time left until next keepalive ping wil be sent to the
```
Agreed, changed to:
```
/** @brief Helper function to determine when next ping should be sent in order
 *	   to keep the connection alive. By default, ping is sent in time
 *	   automatically by the library.
 *
```

jtguggedal commented on samples/nrf9160/azure_fota/src/certificates.h:
```diff
@@ -0,0 +1,17 @@
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#define CLOUD_CLIENT_PRIVATE_KEY \
```
Indeed, this is lazy copy-pasting

jtguggedal commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_dps.h:
```diff
@@ -0,0 +1,78 @@
+ *	    connect to an IoT hub.
+ */
+int dps_start(void);
+
```
Documentation added. But I don't think it should be added to the actual library docs generated by doxygen and published, as it's only used internally in the library

jtguggedal commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_cloud.c:
```diff
@@ -0,0 +1,187 @@
+/*
```
Indeed, I'll change it to your suggestion

rlubos commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1322 @@
+
+		azure_iot_hub_notify_event(&evt);
+		break;
+	case MQTT_EVT_PUBLISH: {
```
nit: I don't see a reason to add braces here (would be consistent with the rest of the code).

jtguggedal replied:
Fixed, not needed after stuff was moved into `on_publish()`

rlubos commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1322 @@
+			LOG_DBG("IPv4 Address found %s", log_strdup(ipv4_addr));
+			break;
+		} else if ((addr->ai_addrlen == sizeof(struct sockaddr_in6)) &&
+			   (AZURE_IOT_HUB_AF_FAMILY == AF_INET6)) {
```
`AZURE_IOT_HUB_AF_FAMILY` is statically assigned to AF_INET, are there plans to make it configurable in the future?

jtguggedal replied:
Removed IPv6 stuff entirely, it's currently not supported by Azure IoT Hub, so it's just noise in the code

rlubos commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1322 @@
+
+			inet_ntop(AF_INET6, &broker6->sin6_addr.s6_addr,
+				  ipv6_addr, sizeof(ipv6_addr));
+			LOG_DBG("IPv4 Address found %s", log_strdup(ipv6_addr));
```
s/IPv4/IPv6

rlubos commented on samples/nrf9160/azure_iot_hub/prj.conf:
```diff
@@ -0,0 +1,65 @@
+CONFIG_MAIN_STACK_SIZE=16384
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=8192
+
+CONFIG_NO_OPTIMIZATIONS=y
```
Is it intended to build the sample w/o optimizations or is it some leftover?

jtguggedal replied:
Debug leftovers, removed here and other places

rlubos commented on samples/nrf9160/azure_iot_hub/prj.conf:
```diff
@@ -0,0 +1,65 @@
+# Heap and stacks
+CONFIG_HEAP_MEM_POOL_SIZE=4096
+CONFIG_MAIN_STACK_SIZE=16384
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=8192
```
Do we need such a large workqueue stack? Neither Azure libs nor the sample make use of the system workqueue.
16k of main stack, it's also quite much, is it really needed?

jtguggedal replied:
Nope, not needed at all. I ran the thread analyzer and adjusted all stacks in both samples

rlubos commented on samples/nrf9160/azure_iot_hub/prj_qemu_x86.conf:
```diff
@@ -0,0 +1,89 @@
+CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y
+CONFIG_FPU=y
+CONFIG_SSE=y
+CONFIG_IDLE_STACK_SIZE=16384
```
Why is this needed?

jtguggedal replied:
Removed

rlubos replied:
Still present

rlubos commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,458 @@
+#include <stdlib.h>
+#include <net/azure_iot_hub.h>
+#include <dk_buttons_and_leds.h>
+#include <dfu/mcuboot.h>
```
Is mcuboot header needed here, I don't see where it could be used.

rlubos commented on samples/nrf9160/azure_fota/prj.conf:
```diff
@@ -0,0 +1,90 @@
+CONFIG_MAIN_STACK_SIZE=8192
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=4096
+
+CONFIG_NO_OPTIMIZATIONS=y
```
Same as in the other sample, is this intended?

rlubos commented on samples/nrf9160/azure_fota/prj.conf:
```diff
@@ -0,0 +1,90 @@
+# Heap and stacks
+CONFIG_HEAP_MEM_POOL_SIZE=4096
+CONFIG_MAIN_STACK_SIZE=8192
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=4096
```
Plus again, I'd check if we really need to increase system workqueue size.

rlubos reviewed (commented):
Massive, but looks pretty good.

jtguggedal commented on subsys/net/lib/azure_fota/azure_fota.c:
```diff
@@ -0,0 +1,555 @@
+		/* Only if CONFIG_FOTA_DOWNLOAD_PROGRESS_EVT is enabled */
+		download_progress = fota_evt->progress;
+		evt.type = AZURE_FOTA_EVT_DL_PROGRESS;
+		evt.dl.progress = download_progress;
```
No, the events are so frequent and end up spamming the system without providing much value. I don't really see the use for these events, so I think I'll just remove them

jtguggedal commented:
@rlubos Thanks for reviewing. Comments have been adressed and rebase done after the upmerge

rlubos commented on samples/nrf9160/azure_iot_hub/prj.conf:
```diff
@@ -0,0 +1,62 @@
+
+# Heap and stacks
+CONFIG_HEAP_MEM_POOL_SIZE=4096
+# CONFIG_MAIN_STACK_SIZE=16384
```
I'd remove this unused config.

jtguggedal replied:
Fixed

rlubos reviewed (approved):
Thanks, no more comments from my side.

simensrostad commented on include/net/azure_fota.h:
```diff
@@ -0,0 +1,91 @@
+#endif
+
+enum azure_fota_evt_type {
+	/** Azure FOTA state report is created and ready to be sent. */
```
nit: Punctuation mark not consistent with the rest of the event descriptions.

simensrostad commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+ *
+ *  @return 0 If successful. Otherwise, a (negative) error code is returned.
+ */
+int azure_iot_hub_init(const struct azure_iot_hub_config *config,
```
It would be nice if return error codes could be listed in the APIs of this PR. This is especially nice for functions returning several different error codes, to allow for faster debugging, and should be common practice for APIs. But hey, that's my opinion.

jtguggedal replied:
Agreed on that it's convenient to list all codes. It's not done here because error codes from other libraries/drivers are propagated and not translated to codes specific to the function, which makes it hard to list all possible codes

simensrostad commented on samples/nrf9160/azure_fota/src/main.c:
```diff
@@ -0,0 +1,248 @@
+	}
+	case LTE_LC_EVT_RRC_UPDATE:
+		printk("RRC mode: %s\n",
+			evt->rrc_mode == LTE_LC_RRC_MODE_CONNECTED ?
```
nit: Alignment

simensrostad commented on samples/nrf9160/azure_fota/src/main.c:
```diff
@@ -0,0 +1,248 @@
+		break;
+	case LTE_LC_EVT_CELL_UPDATE:
+		printk("LTE cell changed: Cell ID: %d, Tracking area: %d\n",
+			evt->cell.id, evt->cell.tac);
```
nittery: alignmento

simensrostad commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,460 @@
+		 * IoT Hub connection is established and interaction with the
+		 * cloud can begin.
+		 *
+		 * The below work submission will  cause send_event() to be
```
nit: extra space between will and cause

simensrostad commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,460 @@
+		.rid = method_data.request_id,
+		.status = 200,
+		.payload = ret,
+		.payload_len = sizeof(ret) - 1,
```
nit: maybe use strlen() and remove the -1. Since the payload is a null-terminated string. Includes less magic.

jtguggedal replied:
I can change it, but usually before the compile time `sizeof` for constants

simensrostad commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,460 @@
+	switch (evt->type) {
+	case LTE_LC_EVT_NW_REG_STATUS:
+		if ((evt->nw_reg_status != LTE_LC_NW_REG_REGISTERED_HOME) &&
+		     (evt->nw_reg_status != LTE_LC_NW_REG_REGISTERED_ROAMING)) {
```
Alignment.

simensrostad commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,460 @@
+		}
+
+		printk("Network registration status: %s\n",
+			evt->nw_reg_status == LTE_LC_NW_REG_REGISTERED_HOME ?
```
And here.

simensrostad commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,460 @@
+		break;
+	case LTE_LC_EVT_PSM_UPDATE:
+		printk("PSM parameter update: TAU: %d, Active time: %d\n",
+			evt->psm_cfg.tau, evt->psm_cfg.active_time);
```
Alignment.

simensrostad commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,460 @@
+	}
+	case LTE_LC_EVT_RRC_UPDATE:
+		printk("RRC mode: %s\n",
+			evt->rrc_mode == LTE_LC_RRC_MODE_CONNECTED ?
```
Here

simensrostad commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,460 @@
+		break;
+	case LTE_LC_EVT_CELL_UPDATE:
+		printk("LTE cell changed: Cell ID: %d, Tracking area: %d\n",
+			evt->cell.id, evt->cell.tac);
```
Hereinho

simensrostad commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,460 @@
+
+		if (err) {
+			printk("Modem could not be configured, error: %d\n",
+				err);
```
Alignment.

simensrostad commented on subsys/net/lib/azure_fota/azure_fota.c:
```diff
@@ -0,0 +1,546 @@
+
+	switch (fota_evt->id) {
+	case FOTA_DOWNLOAD_EVT_FINISHED:
+		LOG_INF("FOTA download completed evt received");
```
Could write without abbreviations here (evt). Alternatively, log the event names as they get propagated.

jtguggedal replied:
I reduced the log level to debug, and left the content as is. I think it's okay to keep debug level stuff abbreviated

simensrostad commented on subsys/net/lib/azure_fota/azure_fota.c:
```diff
@@ -0,0 +1,546 @@
+
+	cJSON_Init();
+
+	/* Get current firmware information, report previous update? */
```
Is this comment debug residue?

simensrostad commented on subsys/net/lib/azure_fota/azure_fota.c:
```diff
@@ -0,0 +1,546 @@
+
+	LOG_DBG("Current version: %s", log_strdup(get_current_version()));
+
+	return 0;
```
Could return err here.

jtguggedal replied:
I think returning `0` makes it a bit clearer that twe would never reach here if an error occured, and an extra bonus is a tighter scope of `err` and save one load from memory

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_cloud_api.c:
```diff
@@ -0,0 +1,192 @@
+		cloud_evt.data.msg.endpoint.type = CLOUD_EP_TOPIC_MSG;
+		cloud_evt.data.msg.endpoint.str = evt->topic.str;
+		cloud_evt.data.msg.endpoint.len = evt->topic.len;
+
```
nit: Unnecessary space, not consistent with the other events.

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+ */
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
+#define USER_NAME_BUF_LEN	(CONFIG_AZURE_IOT_HUB_HOSTNAME_MAX_LEN + \
+				CONFIG_AZURE_IOT_HUB_DEVICE_ID_MAX_LEN + \
```
Alignment?

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+/* Static functions */
+static void azure_iot_hub_notify_event(struct azure_iot_hub_evt *evt)
+{
+	evt_handler(evt);
```
Probably handled somewhere else in the library, but you could add a check here to make sure the handler is not null.

jtguggedal replied:
It should not be possible for events to happen if `evt_handler == NULL`, but I'll add a check anyway

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+	char *ptr;
+
+	/* Topic format:
+	 *   $iothub/twin/res/{status}/?$rid={request id}&$version={version}
```
nit: alignment. Maybe it is intended, Looks good either way.

jtguggedal replied:
It's intended, so I'll leave it as is

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+				  NULL, &hints, &result);
+		if (err) {
+			LOG_ERR("getaddrinfo, error %d", err);
+			return -ECHILD;
```
I'm not sure what the "no children" error codes mean. Doesn't seem to fit here?

jtguggedal replied:
Looks like it's commonly used in NCS, at least in all cloud backends, but can't see any consistency across the rest of the codebase. Anyway, I'll switch this to return the error code from `getaddrinfo`.

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+
+	if (err) {
+		LOG_ERR("getaddrinfo, error %d", err);
+		return -ECHILD;
```
Here too.

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+	}
+
+	err = tls_credential_add(CONFIG_AZURE_IOT_HUB_SEC_TAG,
+					TLS_CREDENTIAL_PRIVATE_KEY,
```
The alignment is off here.

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+	}
+
+	err = tls_credential_add(CONFIG_AZURE_IOT_HUB_SEC_TAG,
+					TLS_CREDENTIAL_CA_CERTIFICATE,
```
Alignment.

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+
+	k_sem_give(&connection_poll_sem);
+
+	return 0;
```
Could return err.

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+		break;
+	case AZURE_IOT_HUB_TOPIC_TWIN_REPORTED:
+		len = snprintk(topic, sizeof(topic),
+				TOPIC_TWIN_REPORT, k_uptime_get_32());
```
Alignment

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+	switch (tx_data->topic.type) {
+	case AZURE_IOT_HUB_TOPIC_EVENT:
+		len = snprintk(topic, sizeof(topic),
+				TOPIC_EVENTS, conn_config.device_id);
```
Alignment

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+
+	connection_state = STATE_INIT;
+
+	k_sleep(K_MSEC(100));
```
Why sleep here?

jtguggedal replied:
Can't recall, probably some debug leftovers, will remove

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+
+	evt_handler = event_handler;
+	is_initialized = true;
+	(void)err;
```
why is err void casted here?

jtguggedal replied:
That's because if DPS and FOTA are disabled, it will otherwise give an "unused variable" compiler warning

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+
+	connection_state = STATE_INIT;
+
+	return 0;
```
Could return err?

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:
```diff
@@ -0,0 +1,612 @@
+
+	LOG_DBG("Azure IoT Hub acquired using DPS was successfully saved");
+
+	return 0;
```
Could return err here. Consistent with ```dps_settings_init```

jtguggedal replied:
Same reasoning as earlier, changed `dps_settings_init` to align

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:
```diff
@@ -0,0 +1,612 @@
+
+	dps_reg_status.state = DPS_STATE_REGISTERING;
+
+	return 0;
```
Could return err.

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:
```diff
@@ -0,0 +1,612 @@
+
+	LOG_DBG("Successfully subscribed to DPS topics");
+
+	return 0;
```
Here too?

simensrostad reviewed (commented):
Well done! I have left a few comments to consider.

jtguggedal commented on samples/nrf9160/azure_fota/src/main.c:
```diff
@@ -0,0 +1,248 @@
+	/* All initializations were successful mark image as working so that we
+	 * will not revert upon reboot.
+	 */
+	boot_write_img_confirmed();
```
I'll leave it here so that it's easy to spot for users of the sample, but it now waits on a semaphore that's given in the `AZURE_IOT_HUB_READY_EVENT` case.

jtguggedal commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -0,0 +1,1303 @@
+
+/* @brief Get report result's status code.
+ *
+ * @param status Pointer to status.
```
Removed doxygen comment style on this static function and simplified it a bit

jtguggedal commented on samples/nrf9160/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,460 @@
+	static const char ret[] = "{\"it\":\"worked\"}";
+	struct azure_iot_hub_result result = {
+		.rid = method_data.request_id,
+		.status = 200,
```
Added comment to explain the result code

jtguggedal commented:
@simensrostad I _think_ I have addressed all your comments now

Also rebased on master branch (97117310ec6b499e86fb4ea9e59b871c251e3da8) and squashed in a doc review commit for both the libraries.

umapraseeda commented:
Adding the latest documentation preview..(Does not include the latest suggestions by @simensrostad as its still being worked on) : [Azure FOTA — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5188079/Azure.FOTA.nRF.Connect.SDK.1.3.99.documentation.pdf)
[Azure IoT Hub — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5188080/Azure.IoT.Hub.nRF.Connect.SDK.1.3.99.documentation.pdf)
[nRF9160_ Azure FOTA — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5188081/nRF9160_.Azure.FOTA.nRF.Connect.SDK.1.3.99.documentation.pdf)
[nRF9160_ Azure IoT Hub — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5188082/nRF9160_.Azure.IoT.Hub.nRF.Connect.SDK.1.3.99.documentation.pdf)

coderbyheart commented on include/net/azure_fota.rst:
```diff
@@ -0,0 +1,72 @@
+			"host": "<hostname>",
+			"path": "<filename>.bin"
+		},
+		"fwFragmentSize" : 1800
```
```suggestion
		"fwFragmentSize": 1800
```

coderbyheart replied:
Not fixed.

jtguggedal replied:
Should be okay now (also the other occurence)

coderbyheart commented on include/net/azure_fota.rst:
```diff
@@ -0,0 +1,72 @@
+Currently, a new version of the firmware is downloaded and applied only if the version is different from the current firmware version.
+
+Currently, the library does not use the ``protocol`` field.
+Instead, the :option:`CONFIG_AZURE_FOTA_TLS` option is used at compile-time to specify the transport protocol to be used.
```
```suggestion
Instead, the :option:`CONFIG_AZURE_FOTA_TLS` option is used at compile-time to specify if HTTPs should be used as the transport protocol.
```

This is a boolean switch, so the protocol cannot be configured.

coderbyheart commented on include/net/azure_fota.rst:
```diff
@@ -0,0 +1,72 @@
+* With TLS: 2 KB including the HTTP header
+* Without TLS: 4 KB
+
+``fwFragmentSize`` should therefore be set lower than the maximum buffer size to keep space for the HTTP header also.
```
```suggestion
``fwFragmentSize`` should therefore be set to a value lower than the maximum buffer size to reserve space for the HTTP header.
```

coderbyheart commented on include/net/azure_fota.h:
```diff
@@ -0,0 +1,91 @@
+ *  @param size Size of buffer.
+ *
+ *  @retval 0 If no FOTA information found.
+ *	    1 If FOTA information was processed.
```
```suggestion
 *  @retval 1 If FOTA information was processed.
```

jtguggedal replied:
Nice catch

coderbyheart commented on include/net/azure_fota.h:
```diff
@@ -0,0 +1,91 @@
+struct azure_fota_event {
+	/** Event type, @ref enum azure_fota_evt_type. */
+	enum azure_fota_evt_type type;
+	/** Pointer to report associated with the event. A report is a JSON
```
```suggestion
	/** Pointer to the report associated with the event. A report is a JSON
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,129 @@
+
+.. _azure_iot_hub_flash_certs:
+
+Provisioning the certificates
```
```suggestion
Provisioning of the certificates
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,129 @@
+	      .device_id_len = sizeof("my-device") - 1,
+	   };
+
+            err = azure_iot_hub_init(&cfg, event_handler);
+            if (err) {
+               printk("azure_iot_hub_init failed: %d\n", err);
+	       return err;
+            }
```
```suggestion
           err = azure_iot_hub_init(&cfg, event_handler);
           if (err) {
              printk("azure_iot_hub_init failed: %d\n", err);
	      return err;
           }
```

coderbyheart commented on samples/nrf9160/azure_fota/README.rst:
```diff
@@ -0,0 +1,188 @@
+	            "host": "my-storage-account.blob.core.windows.net",
+	            "path": "my-app-update.bin"
+	        },
+	        "fwFragmentSize" : 1800
```
```suggestion
	        "fwFragmentSize": 1800
```

umapraseeda commented:
Latest pdfs
[Azure FOTA — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5194520/Azure.FOTA.nRF.Connect.SDK.1.3.99.documentation.pdf)
[Azure IoT Hub — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5194521/Azure.IoT.Hub.nRF.Connect.SDK.1.3.99.documentation.pdf)
[nRF9160_ Azure FOTA — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5194522/nRF9160_.Azure.FOTA.nRF.Connect.SDK.1.3.99.documentation.pdf)
[nRF9160_ Azure IoT Hub — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5194523/nRF9160_.Azure.IoT.Hub.nRF.Connect.SDK.1.3.99.documentation.pdf)

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+	AZURE_IOT_HUB_EVT_DPS_CONNECTING,
+	/** Device Provisioning Service registration has started. */
+	AZURE_IOT_HUB_EVT_DPS_REGISTERING,
+	/** Device Provisioning Service is done, and IoT hub information
```
```suggestion
	/** Device Provisioning Service is done, and Azure IoT Hub information
```

coderbyheart commented on doc/nrf/links.txt:
```diff
@@ -239,6 +239,23 @@
+.. _`Azure IoT Hub device twin messages`: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-device-twins
+.. _`Device enrollments with Azure Portal`: https://docs.microsoft.com/en-us/azure/iot-dps/how-to-manage-enrollments
+.. _`Add certificates to the DPS instance`: https://docs.microsoft.com/en-us/azure/iot-dps/how-to-verify-certificates
+.. _`Registering the device with the IoT hub`: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-create-through-portal#register-a-new-device-in-the-iot-hub
```
```suggestion
.. _`Registering the device with the Azure IoT Hub`: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-create-through-portal#register-a-new-device-in-the-iot-hub
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+enum azure_iot_topic_type {
+	/** Data received on the devicebound topic. */
+	AZURE_IOT_HUB_TOPIC_DEVICEBOUND,
+	/** Event topic used to send event data to IoT hub. */
```
```suggestion
	/** Event topic used to send event data to Azure IoT Hub. */
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,148 @@
+.. note::
+   While `Registering the device with the IoT hub`_, make sure that you select ``X.509 CA Signed`` as the *Authentication type*.
+
+The connection to Azure IoT Hub using MQTT is secured using TLS.
```
```suggestion
The connection to Azure IoT Hub with MQTT is secured using TLS.
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,148 @@
+   While `Registering the device with the IoT hub`_, make sure that you select ``X.509 CA Signed`` as the *Authentication type*.
+
+The connection to Azure IoT Hub using MQTT is secured using TLS.
+For testing purposes, see `Creating Azure IoT Hub certificates`_ for the steps to create certificates and private key for the leaf device, and to register the generated test root certificate to be used with an IoT hub.
```
```suggestion
For testing purposes, see `Creating Azure IoT Hub certificates`_ for the steps to create certificates and a private key for the leaf device, and to register the generated test root certificate to be used with an IoT hub.
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,148 @@
+The connection to Azure IoT Hub using MQTT is secured using TLS.
+For testing purposes, see `Creating Azure IoT Hub certificates`_ for the steps to create certificates and private key for the leaf device, and to register the generated test root certificate to be used with an IoT hub.
+
+The Azure IoT Hub library requires the provisioning of the following certificates and key for a successful TLS connection:
```
```suggestion
The Azure IoT Hub library requires the provisioning of the following certificates and a private key for a successful TLS connection:
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,148 @@
+Provisioning of the certificates
+================================
+
+To provision the certificates to the nRF9160 modem, complete the following steps:
```
```suggestion
To provision the certificates and the private key to the nRF9160 modem, complete the following steps:
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,148 @@
+1. In the `Azure Portal`_, navigate to :guilabel:`IoT Hub` and select the desired IoT hub.
+#. In the overview page, locate and copy the ``Hostname`` and configure :option:`CONFIG_AZURE_IOT_HUB_HOSTNAME` to this address.
+#. Set the option :option:`CONFIG_AZURE_IOT_HUB_DEVICE_ID` to the device ID. The device ID must match the device ID used while creating the certificates.
+#. Set :option:`CONFIG_AZURE_IOT_HUB_SEC_TAG` to the security tag used while :ref:`azure_iot_hub_flash_certs`.
```
```suggestion
#. Set :option:`CONFIG_AZURE_IOT_HUB_SEC_TAG` to the security tag used in :ref:`azure_iot_hub_flash_certs`.
```

coderbyheart commented on include/net/azure_iot_hub.rst:
```diff
@@ -0,0 +1,148 @@
+	      .device_id_len = sizeof("my-device") - 1,
+	  };
+
+          err = azure_iot_hub_init(&cfg, event_handler);
```
```suggestion
          int err = azure_iot_hub_init(&cfg, event_handler);
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+ *		      NULL if CONFIG_AZURE_IOT_HUB_DEVICE_ID is set.
+ *  @param[in] event_handler Pointer to event handler function.
+ *
+ *  @return 0 If successful. Otherwise, a (negative) error code is returned.
```
```suggestion
 *  @retval 0 If successful. Otherwise, a (negative) error code is returned.
```

jtguggedal replied:
My understanding is that `retval` is to be used if an exhaustive list of all possible return values is provided, and `return` in case of a textual description like this.
@ru-fu @umapraseeda Any input on this?

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+ *	   Subsequent calls to other library function should await
+ *	   AZURE_IOT_HUB_EVT_CONNECTED and AZURE_IOT_HUB_EVT_READY events.
+ *
+ *  @return 0 If successful.
```
```suggestion
 *  @retval 0 If successful.
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+ *  @param[in] tx_data Pointer to struct containing data to be transmitted to
+ *                     the Azure IoT Hub.
+ *
+ *  @return 0 If successful.
```
```suggestion
 *  @retval 0 If successful.
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+ *	   disconnection procedure, and the event AZURE_IOT_HUB_EVT_DISCONNECTED
+ *	   is received when it is completed.
+ *
+ *  @return 0 If successful.
```
```suggestion
 *  @retval 0 If successful.
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+ *	     but by default this functionality is handled internally in the
+ *	     library.
+ *
+ *  @return 0 If successful.
```
```suggestion
 *  @retval 0 If successful.
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+/** @brief Ping Azure IoT Hub. Must be called periodically to keep connection to
+ *	   broker alive. By default ping is handled internally in the library.
+ *
+ *  @return 0 If successful.
```
```suggestion
 *  @retval 0 If successful.
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+ *  @param[in] result Structure containing result data from the direct method
+ *		      execution.
+ *
+ *  @return 0 If successful.
```
```suggestion
 *  @retval 0 If successful.
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_dps.h:
```diff
@@ -0,0 +1,108 @@
+ *
+ * @param cfg Pointer to DPS configuration structure.
+ *
+ * @return 0 if successul, otherwise a negative error code.
```
```suggestion
 * @retval 0 if successul, otherwise a negative error code.
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_dps.h:
```diff
@@ -0,0 +1,108 @@
+
+/* @brief Send device registration request to Device Provisioning Service.
+ *
+ * @return 0 if successul, otherwise a negative error code.
```
```suggestion
 * @retval 0 if successul, otherwise a negative error code.
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_dps.h:
```diff
@@ -0,0 +1,108 @@
+
+/* @brief Subscribes to DPS related topics.
+ *
+ * @return 0 if successul, otherwise a negative error code.
```
```suggestion
 * @retval 0 if successul, otherwise a negative error code.
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_dps.h:
```diff
@@ -0,0 +1,108 @@
+/* @brief Convenience function to check if DPS registration is currently in
+ *	  progress.
+ *
+ * @return True if DPS registration is in progress, otherwise false.
```
```suggestion
 * @retval True if DPS registration is in progress, otherwise false.
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+ */
+int azure_iot_hub_input(void);
+
+/** @brief Ping Azure IoT Hub. Must be called periodically to keep connection to
+ *	   broker alive. By default ping is handled internally in the library.
```
```suggestion
/** @brief Ping Azure IoT Hub. Must be called periodically to keep the connection
 *	   alive. By default ping is handled internally in the library.
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+struct azure_iot_hub_data {
+	/** Topic data is sent/received on. */
+	struct azure_iot_hub_topic_data topic;
+	/** Pointer to data sent/received from the Azure IoT Hub broker. */
```
```suggestion
	/** Pointer to data sent/received from the Azure IoT Hub. */
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+	size_t payload_len;
+};
+
+/** @brief Struct with data received from Azure IoT Hub broker. */
```
```suggestion
/** @brief Struct with data received from Azure IoT Hub. */
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+ */
+typedef void (*azure_iot_hub_evt_handler_t)(struct azure_iot_hub_evt *evt);
+
+/** @brief Structure for Azure IoT Hub broker connection parameters. */
```
```suggestion
/** @brief Structure for Azure IoT Hub connection parameters. */
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+
+/** @brief Structure for Azure IoT Hub broker connection parameters. */
+struct azure_iot_hub_config {
+	/** Socket for Azure IoT Hub broker connection. Polling on the socket
```
```suggestion
	/** Socket for Azure IoT Hub connection. Polling on the socket
```

coderbyheart commented on include/net/azure_iot_hub.h:
```diff
@@ -0,0 +1,288 @@
+	 *  happens internally in the library.
+	 */
+	int socket;
+	/** Device id for Azure IoT Hub broker connection.
```
```suggestion
	/** Device id for Azure IoT Hub connection.
```

coderbyheart reviewed (commented):
- Please replace all instances of "IoT hub" with "IoT Hub" (note the capital H).

simensrostad reviewed (approved):
LGMT!

coderbyheart commented on samples/nrf9160/azure_iot_hub/README.rst:
```diff
@@ -0,0 +1,148 @@
+   :class: highlight
+
+   {
+     "temperature" : 25.2,
+	"timestamp": 151325
```
```suggestion
     "temperature": 25.2,
     "timestamp": 151325
```

coderbyheart commented on samples/nrf9160/azure_iot_hub/README.rst:
```diff
@@ -0,0 +1,148 @@
+#. Reset the development kit.
+#. Observe the log output and verify that it is similar to the :ref:`sampoutput_azure_iot`.
+#. Use the `Azure IoT Explorer`_, or log in to the `Azure Portal`_.
+#. Select the connected IoT hub and IoT device.
```
```suggestion
#. Select the connected IoT hub and then your device.
```

jtguggedal commented:
@coderbyheart I've implemented your suggested changes, except the ones with changing `return` -> `retval`, as I'm not quite sure how to interpret the docs: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#doxygen-sg

coderbyheart commented:
> @coderbyheart I've implemented your suggested changes, except the ones with changing `return` -> `retval`, as I'm not quite sure how to interpret the docs: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#doxygen-sg

If you can specify an exact return value, use `@retval`, like you do here:

https://github.com/nrfconnect/sdk-nrf/pull/2715/files#diff-b2074511f589622e6e5fe27e163ad7b8R66

ru-fu commented:
> > @coderbyheart I've implemented your suggested changes, except the ones with changing `return` -> `retval`, as I'm not quite sure how to interpret the docs: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#doxygen-sg
> 
> If you can specify an exact return value, use `@retval`, like you do here:
> 
> https://github.com/nrfconnect/sdk-nrf/pull/2715/files#diff-b2074511f589622e6e5fe27e163ad7b8R66

Yes, exactly - @retval should be used for specific return values, while @return should be used for a general explanation for what is returned. Or, to say it more technically, @retval takes two arguments (value + explanation), while @return takes only one (explanation). 
It's quite easy to see in the output. See for example https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/bl_storage.html , which has some functions with return values and some with just return.

jtguggedal commented:
> > > @coderbyheart I've implemented your suggested changes, except the ones with changing `return` -> `retval`, as I'm not quite sure how to interpret the docs: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#doxygen-sg
> > 
> > 
> > If you can specify an exact return value, use `@retval`, like you do here:
> > https://github.com/nrfconnect/sdk-nrf/pull/2715/files#diff-b2074511f589622e6e5fe27e163ad7b8R66
> 
> Yes, exactly - @RetVal should be used for specific return values, while @return should be used for a general explanation for what is returned. Or, to say it more technically, @RetVal takes two arguments (value + explanation), while @return takes only one (explanation).
> It's quite easy to see in the output. See for example https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/bl_storage.html , which has some functions with return values and some with just return.

So for this phrase that's copy/pasted all over in NCS (at least by me..):

```
0 If successful.
Otherwise, a negative value is returned.
```

What's correct use here, `retval` or `return`? `Otherwise, a negative value is returned.` doesn't list all possible return values, as it's often propagated from other function calls within the function being documented, making it pretty much impossible to list all codes that can be returned.

ru-fu commented:
> So for this phrase that's copy/pasted all over in NCS (at least by me..):
> 
> ```
> 0 If successful.
> Otherwise, a negative value is returned.
> ```
> 
> What's correct use here, `retval` or `return`? `Otherwise, a negative value is returned.` doesn't list all possible return values, as it's often propagated from other function calls within the function being documented, making it pretty much impossible to list all codes that can be returned.

These cases are tricky of course, but we should use ``@retval`` and put the "otherwise" information in the explanation, as it is done in the copy/paste. ;)

It's not really the most correct way to do it, that would be "@retval 0 If successful. @return A negative value if not successful." - but this way, the two lines would show up in different places, and that just makes it really difficult to understand.

jtguggedal commented:
> > So for this phrase that's copy/pasted all over in NCS (at least by me..):
> > ```
> > 0 If successful.
> > Otherwise, a negative value is returned.
> > ```
> > 
> > 
> > What's correct use here, `retval` or `return`? `Otherwise, a negative value is returned.` doesn't list all possible return values, as it's often propagated from other function calls within the function being documented, making it pretty much impossible to list all codes that can be returned.
> 
> These cases are tricky of course, but we should use `@retval` and put the "otherwise" information in the explanation, as it is done in the copy/paste. ;)
> 
> It's not really the most correct way to do it, that would be "@RetVal 0 If successful. @return A negative value if not successful." - but this way, the two lines would show up in different places, and that just makes it really difficult to understand.

@ru-fu Thanks for clearing that up! I'll apply @coderbyheart's suggested changes to keep it in line. 

(I think a clean-up in NCS on this stuff might be needed, I couldn't find much consistency on Doxygen documentation last time I checked)

jtguggedal commented:
@coderbyheart All suggested changes to doxygen comments have now been squashed in

coderbyheart commented:
@umapraseeda can you provide the updated PDFs, please?

umapraseeda commented:
[Azure FOTA — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5200466/Azure.FOTA.nRF.Connect.SDK.1.3.99.documentation.pdf)
[Azure IoT Hub — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5200467/Azure.IoT.Hub.nRF.Connect.SDK.1.3.99.documentation.pdf)
[nRF9160_ Azure FOTA — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5200468/nRF9160_.Azure.FOTA.nRF.Connect.SDK.1.3.99.documentation.pdf)
[nRF9160_ Azure IoT Hub — nRF Connect SDK 1.3.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrf/files/5200469/nRF9160_.Azure.IoT.Hub.nRF.Connect.SDK.1.3.99.documentation.pdf)

umapraseeda reviewed (approved):
LGTM

coderbyheart reviewed (approved):
This is ready to be released! :rocket: :crossed_fingers: 

I did run the sample against a fresh IoT hub and everything worked as expected.

I failed the first time, because I followed this guide: https://github.com/Azure/azure-iot-sdk-c/blob/master/tools/CACertificates/CACertificateOverview.md, however the Client certificate is only `new-device.cert.pem` and `azure-iot-test-only.intermediate.cert.pem` combined.

<details>
  <summary>Device logs</summary>
        
        *** Booting Zephyr OS build v2.3.0-rc1-ncs1-2407-g91d81dc7e0ce  ***
        Flash regions           Domain          Permissions
        00 00 0x00000 0x08000   Secure          rwxl
        01 31 0x08000 0x100000  Non-Secure      rwxl

        Non-secure callable on 0 placed in flash region 0 with size 32.

        SRAM region             Domain          Permissions
        00 07 0x00000 0x10000   Secure          rwxl
        08 31 0x10000 0x40000   Non-Secure      rwxl

        Peripheral              Domain          Status
        00 NRF_P0              Non-Secure       OK
        01 NRF_CLOCK            Non-Secure      OK
        02 NRF_RTC0             Non-Secure      OK
        03 NRF_RTC1             Non-Secure      OK
        04 NRF_NVMC             Non-Secure      OK
        05 NRF_UARTE1           Non-Secure      OK
        06 NRF_UARTE2           Secure          SKIP
        07 NRF_TWIM2            Non-Secure      OK
        08 NRF_SPIM3            Non-Secure      OK
        09 NRF_TIMER0           Non-Secure      OK
        10 NRF_TIMER1           Non-Secure      OK
        11 NRF_TIMER2           Non-Secure      OK
        12 NRF_SAADC            Non-Secure      OK
        13 NRF_PWM0             Non-Secure      OK
        14 NRF_PWM1             Non-Secure      OK
        15 NRF_PWM2             Non-Secure      OK
        16 NRF_PWM3             Non-Secure      OK
        17 NRF_WDT              Non-Secure      OK
        18 NRF_IPC              Non-Secure      OK
        19 NRF_VMC              Non-Secure      OK
        20 NRF_FPU              Non-Secure      OK
        21 NRF_EGU1            Non-Secure       OK
        22 NRF_EGU2             Non-Secure      OK
        23 NRF_DPPIC            Non-Secure      OK
        24 NRF_GPIOTE1          Non-Secure      OK
        25 NRF_REGULATORS       Non-Secure      OK

        SPM: NS image at 0xc000
        SPM: NS MSP at 0x200273e0
        SPM: NS reset vector at 0x10ec5
        SPM: prepare to jump to Non-Secure image.
        *** Booting Zephyr OS build v2.3.0-rc1-ncs1-2407-g91d81dc7e0ce  ***
        Azure IoT Hub sample started
        Connecting to LTE network
        +CEREG: 2,"0901","0212820C",7,0,0,"11100000","11100000" registration ID: mydevice
        LTE cell changed: Cell ID: 34767372, Tracking area: 230
        PSM parameter update: TAU: -1, Active time: -1
        +CSCON: 1
        RRC mode: Connected
        +CEREG: 5,"0901","0212820C",7,,,"11100000","11100000"
        Network registration status: Connected - roaming
        Connected to LTE network
        AZEIOT_HUB_EVT_DPS_CONNECTING
        AZURE_IOT_HUB_EVT_DPS_REGISTERING
        +CSCON: 0
        RRC mode: Idle
        +CSCON: 1
        RRC mode: Connected
        AZURE_IOT_HUB_EVT_DPS_DONE
        AZURE_IOT_HUBEVT_CONNECTING
        [00:00:19.902,465] <inf> azure_iot_hub_dps: Device ID mydevice is now registered to nrf-sdk.azure-devices.net
        +CSCON: 0
        RRC mode: Idle
        +CSCON: 1
        RRC mode: onnected
        AZURE_IOT_HUB_EVT_CONNECTED
        AZURE_IOT_HUB_EVT_READY
        Sending event:
        {"temperature":25.5,"timestamp":30265}
        Evet wa successfully sent
        Next event will be sent in 20 seconds
        AZURE_IOT_HUB_EVT_TWIN_RECEIVED
        No 'telemetryInterval' objectin the device twin
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.7,"timestamp":50277}
        Eventws successfully sent
        Next event will be sent in 20 seconds
        +CSCON: 1
        RRC mode: Connected
        +CEREG: 5,"0901","021D9516",7,,,"11100000","11100000"
        LTE cell changed: Cell ID: 35493142, Tracking area: 2305
        +CSON 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.7,"timestamp":70297}
        Even was successfully sent
        Next eve will be sent in 20 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle
        +CEREG: 5,"0901","0212820C",7,,,"11100000","11100000"
        LTE cell changed: Cell ID: 34767372, Tracking area: 2305
        Sending event:
        {"temperature":25.7,"timestamp":90317}
        Even wassuccessfully sent
        Next event will be sent in 20 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle
        Senin event:
        {"temperature":25.8,"timestamp":110338}
        Even was successfully sent
        Next event will be sent in 20 seconds
        +CSCON: 1
        RRC mode: Connected
        RRC mode: Idle
        Sending event:
        {"temperature":25.6,"timsamp":130356}
        Event was successfully sent
        Next event will be sent in 20 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.6,"timestamp":150376}
        Event was successfully sent
        Next event will be sent in 20 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.7,"timestamp":170397}
        Event was successfully sent
        Next event will be sent in 20 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.7,"timestamp":190417}
        Event was successfully sent
        Next event will be sent in 20 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.8,"timestamp":210438}
        Event was successfully sent
        Next event will be sent in 20 seconds
        +CSCON: 1
        RRC mode: Connected
        AZURE_IOT_HUB_EVT_TWIN_DESIRED_RECEIVED
        New telemetry interval has been applied: 10
        Sending event:
        {"temperature":25.3,"timestamp":211053}
        Event was successfully sent
        Next event will be sent in 10 seconds
        AZURE_IOT_HUB_EVT_TWIN_RESULT_SUCCESS, ID: 211048
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.5,"timestamp":221065}
        Event was successfully sent
        Next event will be sent in 10 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.6,"timestamp":231086}
        Event was successfully sent
        Next event will be sent in 10 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.6,"timestamp":241106}
        Event was successfully sent
        Next event will be sent in 10 seconds
        +CSCON: 1
        RRC mode: Connected
        AZURE_IOT_HUB_EVT_DIRECT_METHOD
        Method name: led
        Payload: 1
        +CSCON: 0
        RRC mode: Idle
        +CSCON: 1
        RRC mode: Connected
        Sending event:
        {"temperature":25.7,"timestamp":251127}
        Event was successfully sent
        Next event will be sent in 10 seconds
        +CSCON: 0
        RRC mode: Idle
        +CSCON: 1
        RRC mode: Connected
        AZURE_IOT_HUB_EVT_DIRECT_METHOD
        Method name: led
        Payload: 0
        Sending event:
        {"temperature":25.7,"timestamp":261147}
        Event was successfully sent
        Next event will be sent in 10 seconds
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.9,"timestamp":271159}
        Event was successfully sent
        Next event will be sent in 10 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.0,"timestamp":281180}
        Event was successfully sent
        Next event will be sent in 10 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle
        Sending event:
        {"temperature":25.0,"timestamp":291200}
        Event was successfully sent
        Next event will be sent in 10 seconds
        +CSCON: 1
        RRC mode: Connected
        +CSCON: 0
        RRC mode: Idle

</details>

jtguggedal commented:
I had to do minor tweaks to `prj.conf` in both smaples after approvals to make CI build the the samples properly. Diff here: https://github.com/nrfconnect/sdk-nrf/compare/1da76386202c0542a74557e3117713f62dbb483f..915182c25bdf1df1c859c1f6d2b73729e511dbd0

coderbyheart commented:
:rocket:

breadboardbasics commented:
Awesome work on this. I have been working on a project that is on a very old version of the SDK and got Azure working manually which was tricky at the time. Nice to see the development work come this far and support Azure out of the box. 

Quick question: In my code, I am noticing that there seems to be a max of 2 devices connected over MQTT/TLS to Azure at one time. Is this some limitation of the hardware? I notice in the code here that CONFIG_NET_SOCKETS_TLS_MAX_CONTEXTS = 2 which lines up with the behavior I have seen when testing manually. Whenever I try to make a third connection, I get errno -45 which says operation not supported on socket. 

Do you know if it is possible to create more than two simultaneous connections to Azure? From my testing there seems to be some limitation.

coderbyheart commented:
@breadboardbasics please don't use feature PR comments for support questions that are not related to the Azure library. Please head over to [{DevZone](https://devzone.nordicsemi.com/) and create a new support case there.
